grandpa/src/consensus.rs

```diff
@@ -0,0 +1,232 @@
+// Copyright (C) 2023 Polytope Labs.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific lang
+
+use crate::consensus_message::ConsensusMessage;
+use codec::Decode;
+use core::marker::PhantomData;
+use ismp::{
+    consensus::{ConsensusClient, ConsensusStateId, StateCommitment, StateMachineClient},
+    error::Error,
+    host::{IsmpHost, StateMachine},
+    messaging::StateCommitmentHeight,
+};
+use primitive_types::H256;
+use primitives::{ConsensusState, ParachainHeadersWithFinalityProof};
+use sp_consensus_aura::{Slot, AURA_ENGINE_ID};
+use sp_runtime::{traits::Header, DigestItem};
+use std::{collections::BTreeMap, time::Duration};
+use verifier::{
+    verify_grandpa_finality_proof, verify_parachain_headers_with_grandpa_finality_proof,
+};
+
+pub const POLKADOT_CONSENSUS_STATE_ID: [u8; 8] = *b"polkadot";
+pub const KUSAMA_CONSENSUS_STATE_ID: [u8; 8] = *b"_kusama_";
+
+const SLOT_DURATION: u64 = 12_000;
+
+/// The `ConsensusEngineId` of ISMP digest in the parachain header.
+pub const ISMP_ID: sp_runtime::ConsensusEngineId = *b"ISMP";
+
+pub struct GrandpaConsensusClient<T>(PhantomData<T>);
+
+impl<T> Default for GrandpaConsensusClient<T> {
+    fn default() -> Self {
+        Self(PhantomData)
+    }
+}
+
+impl<T> ConsensusClient for GrandpaConsensusClient<T>
+where
+    T: pallet_ismp::Config + super::Config,
+    T::BlockNumber: Into<u32>,
+    T::Hash: From<H256>,
+{
+    fn verify_consensus(
+        &self,
+        host: &dyn IsmpHost,
+        _consensus_state_id: ConsensusStateId,
+        trusted_consensus_state: Vec<u8>,
+        proof: Vec<u8>,
+    ) -> Result<(Vec<u8>, BTreeMap<StateMachine, StateCommitmentHeight>), Error> {
```

Comment:
we might need to modify the return value here in `ismp-rs` support returning multiple state commitments
The update can contain multiple headers for the standalone chain.
```suggestion
    ) -> Result<(Vec<u8>, BTreeMap<StateMachine, Vec<StateCommitmentHeight>>), Error> {
```

Reply:
You can update this now